### ISSUE TITLE
KernelFusion/jit_caching.cpp seen failing on pre-commit HIP CI

### ISSUE THREAD

**Issue by steffenlarsen** [intel/llvm#12096]
The E2E test KernelFusion/jit_caching.cpp was seen fail in pre-commit testing on HIP for unrelated changes. The [failure log](https://github.com/intel/llvm/actions/runs/7095269080/job/19312177469?pr=12071) shows:

```
******************** TEST 'SYCL :: KernelFusion/jit_caching.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 2
/__w/llvm/llvm/toolchain/bin//clang++  -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx10[31](https://github.com/intel/llvm/actions/runs/7095269080/job/19312177469?pr=12071#step:21:32) -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/jit_caching.cpp -fsycl-embed-ir -O2 -o /__w/llvm/llvm/build-e2e/KernelFusion/Output/jit_caching.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/jit_caching.cpp -fsycl-embed-ir -O2 -o /__w/llvm/llvm/build-e2e/KernelFusion/Output/jit_caching.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 3
env SYCL_RT_WARNING_LEVEL=1 env ONEAPI_DEVICE_SELECTOR=ext_oneapi_hip:gpu  /__w/llvm/llvm/build-e2e/KernelFusion/Output/jit_caching.cpp.tmp.out 2>&1 | /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/jit_caching.cpp --implicit-check-not "COMPUTATION ERROR" --implicit-check-not "WRONG INTERNALIZATION"
# executed command: env SYCL_RT_WARNING_LEVEL=1 env ONEAPI_DEVICE_SELECTOR=ext_oneapi_hip:gpu /__w/llvm/llvm/build-e2e/KernelFusion/Output/jit_caching.cpp.tmp.out
# note: command had no output on stdout or stderr
# error: command failed with exit status: -6
# executed command: /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/KernelFusion/jit_caching.cpp --implicit-check-not 'COMPUTATION ERROR' --implicit-check-not 'WRONG INTERNALIZATION'
# note: command had no output on stdout or stderr

--

********************
```

**Comment by steffenlarsen** [comment]
Tag @sommerlukas @jopperm @victor-eds

**Comment by sommerlukas** [comment]
Thanks for reporting @steffenlarsen, we'll look into it.

**Comment by jopperm** [comment]
@steffenlarsen we were unable to reproduce the test crash locally, and it hasn't occurred again in later PRs.

**Comment by steffenlarsen** [comment]
Thank you, @jopperm . Let us keep it open for a another couple of days to see if it happens again in CI. If it doesn't we can write it off as a fluke.

**Comment by jopperm** [comment]
I didn't spot any new occurrences in the PRs that got merged since Friday -- @steffenlarsen shall we close this issue then?

**Comment by steffenlarsen** [comment]
Let's go for it. Closing as unreproducible.

**Comment by againull** [comment]
Unfortunately this error appeared again in unrelated PR https://github.com/intel/llvm/pull/12260

**Comment by jopperm** [comment]
Thanks for reporting, we'll try to get to the bottom of this.

**Comment by jopperm** [comment]
We currently believe that the observed crashes correspond to a failed [assertion](https://github.com/intel/llvm/blob/89327e0a96e9b480e807211be79749741eb78fc1/sycl-fusion/passes/kernel-fusion/SYCLKernelFusion.cpp#L393) in the kernel fusion library. The problem is caused by an out-of-bounds access that sometimes returns seemingly valid data, which could explain the spuriousness of the failures.

The out-of-bounds access is fixed in #12415.

**Comment by jopperm** [comment]
I haven't seen the test failing in the PRs merged after #12415. @againull, @bader are you ok with closing this again?

**Comment by bader** [comment]
👍

**Comment by sommerlukas** [comment]
Closing this now, feel free to reopen or tag us on a new issue if the problem arises again.
